File: problemset/3Sum.cpp
class Solution {
public:
    vector<vector<int>> threeSum(vector<int>& nums) {
      int n = nums.size();
      std::vector<std::vector<int>> ret;
      std::sort( nums.begin(), nums.end() );
      for (int i = 0; i < n && nums[i] <= 0; ++i) {
        if (i && nums[i] == nums[i-1]) continue;
        for (int j = i+1, k = n - 1; j < k; ++j) {
          if (j != i+1 && nums[j] == nums[j-1]) continue;
          if (nums[i] + nums[j] > 0) break;
          int sum = 0;
          while (j < k && ((sum=nums[i] + nums[j] + nums[k]) > 0)) --k;
          if (sum == 0) ret.push_back({nums[i], nums[j], nums[k]});
        }
      }
      return ret;
    }
};

File: activity/20200805/RabbitLee/3Sum.py
class Solution:
    def threeSum(self, nums: List[int]) -> List[List[int]]:
        triplets = []
        nums = sorted(nums)
        
        first = 0
        while first < len(nums) - 2:
            twins = self.twoSum2(nums[first + 1:], -nums[first])
            triplets += [[nums[first]] + twin for twin in twins]
            first += 1
            while first < len(nums) - 2 and nums[first] == nums[first - 1]:
                first += 1
                
        return triplets
    
    def twoSum2(self, numbers: List[int], target: int) -> List[List[int]]:
        twins = list()
        left = 0
        right = len(numbers) - 1
        
        while left < right:
            if numbers[left] + numbers[right] < target:
                left += 1
            elif numbers[left] + numbers[right] > target:
                right -= 1
            else:
                twins.append([numbers[left], numbers[right]])
                right -= 1
                left += 1
                while left < right and numbers[left] == numbers[left - 1]:
                    left += 1
                    
        return twins           

